controller/transfer/association_transfer_manager.py

```diff
@@ -0,0 +1,111 @@
+import traceback
+from typing import Dict, Any
+from controller.labeling_task import manager as labeling_task_manager
+from controller.attribute import manager as attribute_manager
+from controller.record import manager as record_manager
+from util import notification
+
+
+from controller.information_source import manager as information_source_manager
+from submodules.model import enums
+from submodules.model.business_objects import (
+    general,
+    labeling_task_label,
+    record_label_association,
+)
+from submodules.model.models import RecordLabelAssociation
+from controller.weak_supervision import weak_supervision_service as weak_supervision
+
+
+def import_associations(
+    project_id: str,
+    user_id: str,
+    model_name: str,
+    labeling_task_name: str,
+    associations: Dict[str, Any],
+    indices: Dict[str, Any],
+    source_type: str,
+) -> int:
+    labeling_task = labeling_task_manager.get_labeling_task_by_name(
+        project_id, labeling_task_name
+    )
+
+    information_source = information_source_manager.get_information_source_by_name(
+        project_id, model_name
+    )
+    if information_source is None:
+        if source_type == "model_callback":
+            information_source = information_source_manager.create_information_source(
+                project_id,
+                user_id,
+                labeling_task.id,
+                model_name,
+                "",
+                "This is a model callback",
+                enums.LabelSource.MODEL_CALLBACK.value,
+            )
+        elif source_type == "heuristic":
+            information_source = information_source_manager.create_information_source(
+                project_id,
+                user_id,
+                labeling_task.id,
+                model_name,
+                "",
+                "This is a heuristic",
+                enums.LabelSource.INFORMATION_SOURCE.value,
+            )
+        else:
+            raise Exception("Unknown source type")
+
+    attribute_names = list(indices[0].keys())
+    attribute_list = attribute_manager.get_all_attributes_by_names(
+        project_id, attribute_names
+    )
+
+    records = record_manager.get_records_by_composite_keys(
+        project_id,
+        indices,
+        attribute_list,
+        enums.RecordCategory.SCALE.value,  # we currently don't use TEST any more
+    )
+    record_ids = [record.id for record in list(records.values())]
+
+    record_label_associations = []
+    for record_id, association in zip(record_ids, associations):
+        label_name, confidence = association
+
+        label = labeling_task_label.get_by_name(
+            project_id, labeling_task.id, label_name
+        )
```

Comment:
this will run a new query every loop, i'd go with get all, dict comprehension and grabbing the data then.

This would also open the door for list comprehension for the record_label_associations which will struggle with the append on bigger lists.

Also depending on the size it might make sense to batch this a bit sind an insert with > 10k records will run into issues. maybe max 1k records per batch

Reply:
the batch is limited to 32 files per request. I added the change re the label query calling every loop iteration into one before the loop :)